precompiles/conviction-voting/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(assert_matches)]
+
+use fp_evm::PrecompileHandle;
+use frame_support::dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo};
+use frame_support::traits::{Currency, Polling};
+use pallet_conviction_voting::Call as ConvictionVotingCall;
+use pallet_conviction_voting::{AccountVote, Conviction, Tally, Vote};
+use pallet_evm::AddressMapping;
+use precompile_utils::prelude::*;
+use sp_core::{H160, H256, U256};
+use sp_runtime::traits::StaticLookup;
+use sp_std::marker::PhantomData;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+type BalanceOf<Runtime> = <<Runtime as pallet_conviction_voting::Config>::Currency as Currency<
+	<Runtime as frame_system::Config>::AccountId,
+>>::Balance;
+type IndexOf<Runtime> = <<Runtime as pallet_conviction_voting::Config>::Polls as Polling<
+	Tally<
+		<<Runtime as pallet_conviction_voting::Config>::Currency as Currency<
+			<Runtime as frame_system::Config>::AccountId,
+		>>::Balance,
+		<Runtime as pallet_conviction_voting::Config>::MaxTurnout,
+	>,
+>>::Index;
+type ClassOf<Runtime> = <<Runtime as pallet_conviction_voting::Config>::Polls as Polling<
+	Tally<
+		<<Runtime as pallet_conviction_voting::Config>::Currency as Currency<
+			<Runtime as frame_system::Config>::AccountId,
+		>>::Balance,
+		<Runtime as pallet_conviction_voting::Config>::MaxTurnout,
+	>,
+>>::Class;
+
+/// A precompile to wrap the functionality from pallet-conviction-voting.
+pub struct ConvictionVotingPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+impl<Runtime> ConvictionVotingPrecompile<Runtime>
+where
+	Runtime: pallet_conviction_voting::Config + pallet_evm::Config + frame_system::Config,
+	BalanceOf<Runtime>: TryFrom<U256>,
+	<Runtime as frame_system::Config>::Hash: TryFrom<H256>,
+	<Runtime as frame_system::Config>::RuntimeCall:
+		Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<<Runtime as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as frame_system::Config>::RuntimeCall: From<ConvictionVotingCall<Runtime>>,
+	IndexOf<Runtime>: TryFrom<u32>,
+	ClassOf<Runtime>: TryFrom<u16>,
+{
+	/// Vote in a poll.
+	///
+	/// Parameters:
+	/// * poll_index: Index of poll
+	/// * aye: Yes or no vote
+	/// * vote_amount: Balance locked for vote
+	/// * conviction: Conviction multiplier for length of vote lock
+	#[precompile::public("vote(uint256,bool,uint256,uint256)")]
+	fn vote(
+		handle: &mut impl PrecompileHandle,
+		poll_index: SolidityConvert<U256, u32>,
+		aye: bool,
+		vote_amount: U256,
+		conviction: SolidityConvert<U256, u8>,
+	) -> EvmResult {
+		let poll_index = Self::u32_to_index(poll_index.converted()).in_field("poll_index")?;
+		let vote_amount = Self::u256_to_amount(vote_amount).in_field("voteAmount")?;
+		let conviction = Self::u8_to_conviction(conviction.converted()).in_field("conviction")?;
```

Comment:
What does conviction represent in solidity? a multiplier? an enum? Perhaps it would be good if u use a u8 that represents an enum? (similar yo what we use in substrate)

Reply:
An enum, `u8` works but was following the pallet-democracy conventions here https://github.com/PureStake/moonbeam/blob/master/precompiles/pallet-democracy/src/lib.rs#L111

I believe the reason to use U256 is because either way it is U256 in the EVM because this is the smallest variable size. @nanocryk is it something like that

Reply:
The reason I was making the suggestion is because there are other precompiles (like proxy) where the enum is defined in the solidity file, hence developers might have a better notion on what to put: https://github.com/PureStake/moonbeam/blob/6924ede17db074c4d5d530294bc0ff2eb484c674/precompiles/proxy/Proxy.sol#L17